File: consorcio/lib/infrastructure/firestore/firebase_task_repository.dart
import 'dart:async';

import 'package:cloud_firestore/cloud_firestore.dart';

import '../../domain/entities/field_task.dart';
import '../../domain/errors/domain_exception.dart';
import '../../domain/repositories/task_repository.dart';

class FirebaseTaskRepository implements TaskRepository {
  FirebaseTaskRepository({FirebaseFirestore? firestore})
      : _firestore = firestore ?? FirebaseFirestore.instance;

  final FirebaseFirestore _firestore;

  CollectionReference<Map<String, dynamic>> get _tasks =>
      _firestore.collection('tasks');

  double? _coord(dynamic value) {
    if (value is num) {
      final next = value.toDouble();
      return next.isFinite ? next : null;
    }
    if (value is String) {
      final next = double.tryParse(value.trim());
      if (next != null && next.isFinite) return next;
    }
    return null;
  }

  TaskRoute _mapRoute(Map<String, dynamic> data) {
    final latitude = _coord(data['latitude']);
    final longitude = _coord(data['longitude']);
    final hasPoint = latitude != null &&
        longitude != null &&
        !(latitude == 0 && longitude == 0);
    return TaskRoute(
      routeCode: ((data['routeCode'] as String?) ?? '').replaceAll(RegExp(r'\D'), ''),
      latitude: hasPoint ? latitude : null,
      longitude: hasPoint ? longitude : null,
      note: (data['note'] as String?) ?? '',
      completed: data['completed'] == true,
      completedById: (data['completedById'] as String?) ?? '',
      completedByName: (data['completedByName'] as String?) ?? '',
      completedAt: (data['completedAt'] as Timestamp?)?.toDate(),
      claimedById: (data['claimedById'] as String?) ?? '',
      claimedByName: (data['claimedByName'] as String?) ?? '',
      claimedAt: (data['claimedAt'] as Timestamp?)?.toDate(),
      photosUploaded: data['photosUploaded'] == true,
    );
  }

  TaskNotice? _mapNotice(dynamic raw) {
    if (raw is! Map) return null;
    final data = Map<String, dynamic>.from(raw);
    final message = (data['message'] as String?) ?? '';
    if (message.isEmpty) return null;
    return TaskNotice(
      message: message,
      routeCode: (data['routeCode'] as String?) ?? '',
      createdById: (data['createdById'] as String?) ?? '',
      createdByName: (data['createdByName'] as String?) ?? '',
      createdAt: (data['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
    );
  }

  Map<String, dynamic> _serializeRoute(TaskRoute route) {
    final payload = <String, dynamic>{
      'routeCode': route.routeCode,
      'note': route.note,
      'completed': route.completed,
      'completedById': route.completedById,
      'completedByName': route.completedByName,
      'completedAt': route.completedAt == null
          ? null
          : Timestamp.fromDate(route.completedAt!),
      'claimedById': route.claimedById,
      'claimedByName': route.claimedByName,
      'claimedAt': route.claimedAt == null
          ? null
          : Timestamp.fromDate(route.claimedAt!),
      'photosUploaded': route.photosUploaded,
    };
    if (route.hasMapPoint) {
      payload['latitude'] = route.latitude;
      payload['longitude'] = route.longitude;
    }
    return payload;
  }

  FieldTask _map(String id, Map<String, dynamic> data) {
    final rawRoutes = (data['routes'] as List?) ?? const [];
    final routes = rawRoutes
        .whereType<Map>()
        .map((item) => _mapRoute(Map<String, dynamic>.from(item)))
        .where((route) => route.routeCode.isNotEmpty)
        .toList();
    final latitude = _coord(data['latitude']);
    final longitude = _coord(data['longitude']);
    final task = FieldTask(
      id: id,
      title: (data['title'] as String?) ?? '',
      description: (data['description'] as String?) ?? '',
      status: (data['status'] as String?) ?? 'PENDIENTE',
      dueDate: (data['dueDate'] as Timestamp?)?.toDate(),
      areaId: (data['areaId'] as String?) ?? '',
      areaName: (data['areaName'] as String?) ?? '',
      routeCode: (data['routeCode'] as String?) ?? '',
      latitude: latitude,
      longitude: longitude,
      routes: routes,
      lastNotice: _mapNotice(data['lastNotice']),
      assignToAllTechnicians: data['assignToAllTechnicians'] == true,
      assignedTechnicianIds: List<String>.from(
        (data['assignedTechnicianIds'] as List?) ?? const [],
      ),
      assignedTechnicianNames: List<String>.from(
        (data['assignedTechnicianNames'] as List?) ?? const [],
      ),
      createdById: (data['createdById'] as String?) ?? '',
      createdByName: (data['createdByName'] as String?) ?? '',
      completedAt: (data['completedAt'] as Timestamp?)?.toDate(),
      completedById: (data['completedById'] as String?) ?? '',
      completedByName: (data['completedByName'] as String?) ?? '',
      createdAt: (data['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      updatedAt: (data['updatedAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
    );
    return FieldTask(
      id: task.id,
      title: task.title,
      description: task.description,
      status: task.status,
      dueDate: task.dueDate,
      areaId: task.areaId,
      areaName: task.areaName,
      routeCode: task.normalizedRoutes.isEmpty
          ? task.routeCode
          : task.normalizedRoutes.first.routeCode,
      latitude: task.normalizedRoutes.isEmpty
          ? task.latitude
          : task.normalizedRoutes.first.latitude,
      longitude: task.normalizedRoutes.isEmpty
          ? task.longitude
          : task.normalizedRoutes.first.longitude,
      routes: task.normalizedRoutes,
      lastNotice: task.lastNotice,
      assignToAllTechnicians: task.assignToAllTechnicians,
      assignedTechnicianIds: task.assignedTechnicianIds,
      assignedTechnicianNames: task.assignedTechnicianNames,
      createdById: task.createdById,
      createdByName: task.createdByName,
      completedAt: task.completedAt,
      completedById: task.completedById,
      completedByName: task.completedByName,
      createdAt: task.createdAt,
      updatedAt: task.updatedAt,
    );
  }

  List<FieldTask> _merge(Iterable<QueryDocumentSnapshot<Map<String, dynamic>>> docs) {
    final byId = <String, FieldTask>{};
    for (final doc in docs) {
      byId[doc.id] = _map(doc.id, doc.data());
    }
    final list = byId.values.toList()
      ..sort((a, b) => b.updatedAt.compareTo(a.updatedAt));
    return list;
  }

  @override
  Future<List<FieldTask>> listAccessibleForUser(String userId) async {
    final assigned = await _tasks
        .where('assignedTechnicianIds', arrayContains: userId)
        .get();
    final allTechs = await _tasks
        .where('assignToAllTechnicians', isEqualTo: true)
        .get();
    return _merge([...assigned.docs, ...allTechs.docs]);
  }

  @override
  Stream<List<FieldTask>> watchAccessibleForUser(String userId) {
    final controller = StreamController<List<FieldTask>>();
    var assigned = <QueryDocumentSnapshot<Map<String, dynamic>>>[];
    var allTechs = <QueryDocumentSnapshot<Map<String, dynamic>>>[];

    void emit() {
      if (controller.isClosed) return;
      controller.add(_merge([...assigned, ...allTechs]));
    }

    final subAssigned = _tasks
        .where('assignedTechnicianIds', arrayContains: userId)
        .snapshots()
        .listen((snapshot) {
      assigned = snapshot.docs;
      emit();
    }, onError: controller.addError);
    final subAll = _tasks
        .where('assignToAllTechnicians', isEqualTo: true)
        .snapshots()
        .listen((snapshot) {
      allTechs = snapshot.docs;
      emit();
    }, onError: controller.addError);

    controller.onCancel = () async {
      await subAssigned.cancel();
      await subAll.cancel();
    };
    return controller.stream;
  }

  @override
  Future<List<FieldTask>> listAll() async {
    final snapshot = await _tasks.get();
    final list = snapshot.docs
        .map((doc) => _map(doc.id, doc.data()))
        .toList()
      ..sort((a, b) => b.updatedAt.compareTo(a.updatedAt));
    return list;
  }

  @override
  Future<FieldTask> create(CreateFieldTaskInput input) async {
    final id = _firestore.collection('tasks').doc().id;
    final now = Timestamp.now();
    final payload = <String, dynamic>{
      'title': input.title,
      'description': input.description,
      'status': 'PENDIENTE',
      'dueDate':
          input.dueDate == null ? null : Timestamp.fromDate(input.dueDate!),
      'areaId': input.areaId,
      'areaName': input.areaName,
      'routeCode': input.routeCode,
      'routes': input.routes.map(_serializeRoute).toList(),
      'assignToAllTechnicians': input.assignToAllTechnicians,
      'assignedTechnicianIds': input.assignedTechnicianIds,
      'assignedTechnicianNames': input.assignedTechnicianNames,
      'createdById': input.createdById,
      'createdByName': input.createdByName,
      'completedAt': null,
      'completedById': '',
      'completedByName': '',
      'createdAt': now,
      'updatedAt': now,
    };
    if (input.latitude != null && input.longitude != null) {
      payload['latitude'] = input.latitude;
      payload['longitude'] = input.longitude;
    }
    await _tasks.doc(id).set(payload);
    return _map(id, payload);
  }

  @override
  Future<FieldTask?> getById(String id) async {
    final snapshot = await _tasks.doc(id).get();
    if (!snapshot.exists || snapshot.data() == null) return null;
    return _map(snapshot.id, snapshot.data()!);
  }

  @override
  Future<FieldTask> update(String id, UpdateFieldTaskInput input) async {
    final ref = _tasks.doc(id);
    final snapshot = await ref.get();
    if (!snapshot.exists || snapshot.data() == null) {
      throw DomainException('Tarea no encontrada');
    }
    final current = _map(id, snapshot.data()!);
    final patch = _buildPatch(current, input);
    await _commitPatch(ref, patch);
    return _map(id, {...snapshot.data()!, ...patch});
  }

  @override
  Future<FieldTask> transact(
    String id,
    UpdateFieldTaskInput Function(FieldTask current) builder,
  ) async {
    final ref = _tasks.doc(id);
    try {
      return await _firestore.runTransaction((transaction) async {
        final snapshot = await transaction.get(ref);
        if (!snapshot.exists || snapshot.data() == null) {
          throw DomainException('Tarea no encontrada');
        }
        final current = _map(id, snapshot.data()!);
        final input = builder(current);
        final patch = _buildPatch(current, input);
        transaction.update(ref, patch);
        return _map(id, {...snapshot.data()!, ...patch});
      });
    } on DomainException {
      rethrow;
    } on FirebaseException catch (error) {
      throw _mapWriteError(error);
    }
  }

  Map<String, dynamic> _buildPatch(
    FieldTask current,
    UpdateFieldTaskInput input,
  ) {
    final routes = input.routes ?? current.normalizedRoutes;
    final primary = routes.isEmpty ? null : routes.first;
    final patch = <String, dynamic>{
      if (input.status != null) 'status': input.status,
      'routes': routes.map(_serializeRoute).toList(),
      'updatedAt': Timestamp.now(),
    };

    if (primary != null && primary.hasMapPoint) {
      patch['latitude'] = primary.latitude;
      patch['longitude'] = primary.longitude;
    }

    if (input.lastNotice != null) {
      patch['lastNotice'] = {
        'message': input.lastNotice!.message,
        'routeCode': input.lastNotice!.routeCode,
        'createdById': input.lastNotice!.createdById,
        'createdByName': input.lastNotice!.createdByName,
        'createdAt': Timestamp.fromDate(input.lastNotice!.createdAt),
      };
    }
    if (input.clearCompletedAt) {
      patch['completedAt'] = null;
      patch['completedById'] = '';
      patch['completedByName'] = '';
    } else {
      if (input.completedAt != null) {
        patch['completedAt'] = Timestamp.fromDate(input.completedAt!);
      }
      if (input.completedById != null && input.completedById!.isNotEmpty) {
        patch['completedById'] = input.completedById;
      }
      if (input.completedByName != null &&
          input.completedByName!.isNotEmpty) {
        patch['completedByName'] = input.completedByName;
      }
    }
    return patch;
  }

  Future<void> _commitPatch(
    DocumentReference<Map<String, dynamic>> ref,
    Map<String, dynamic> patch,
  ) async {
    try {
      await ref.update(patch);
    } on FirebaseException catch (error) {
      throw _mapWriteError(error);
    }
  }

  Never _mapWriteError(FirebaseException error) {
    if (error.code == 'permission-denied') {
      throw DomainException(
        'No se pudo guardar el avance. Vuelve a intentar.',
      );
    }
    throw DomainException('No se pudo actualizar la tarea');
  }
}

File: src/infrastructure/firestore/FirebaseTaskRepository.ts
import {
  collection,
  deleteDoc,
  doc,
  getDoc,
  getDocs,
  onSnapshot,
  orderBy,
  query,
  setDoc,
  Timestamp,
  updateDoc,
  where,
} from 'firebase/firestore'
import type { Task, TaskNotice, TaskRoute, TaskStatus } from '@/domain/entities/Task'
import {
  isValidMapCoord,
  normalizeTaskRoutes,
  primaryTaskRoute,
  TaskStatus as Status,
} from '@/domain/entities/Task'
import type {
  CreateTaskInput,
  TaskRepository,
  UpdateTaskInput,
} from '@/domain/repositories/TaskRepository'
import { NotFoundError } from '@/domain/errors/DomainError'
import { firestoreDb } from '@/infrastructure/firebase/firebaseApp'

interface TaskRouteDoc {
  routeCode: string
  latitude?: number
  longitude?: number
  note?: string
  completed?: boolean
  completedById?: string
  completedByName?: string
  completedAt?: Timestamp | null
  claimedById?: string
  claimedByName?: string
  claimedAt?: Timestamp | null
  photosUploaded?: boolean
}

interface TaskNoticeDoc {
  message: string
  routeCode: string
  createdById: string
  createdByName: string
  createdAt: Timestamp
}

interface TaskDoc {
  title: string
  description: string
  status: TaskStatus
  dueDate: Timestamp | null
  areaId: string
  areaName: string
  routeCode: string
  latitude?: number
  longitude?: number
  routes?: TaskRouteDoc[]
  lastNotice?: TaskNoticeDoc | null
  assignToAllTechnicians: boolean
  assignedTechnicianIds: string[]
  assignedTechnicianNames: string[]
  createdById: string
  createdByName: string
  completedAt: Timestamp | null
  completedById: string
  completedByName: string
  createdAt: Timestamp
  updatedAt: Timestamp
}

function mapRoute(data: TaskRouteDoc): TaskRoute {
  const latitude =
    typeof data.latitude === 'number' && Number.isFinite(data.latitude)
      ? data.latitude
      : null
  const longitude =
    typeof data.longitude === 'number' && Number.isFinite(data.longitude)
      ? data.longitude
      : null
  const hasCoords = isValidMapCoord(latitude, longitude)
  return {
    routeCode: String(data.routeCode ?? '').replace(/\D/g, ''),
    latitude: hasCoords ? latitude : null,
    longitude: hasCoords ? longitude : null,
    note: (data.note ?? '').trim(),
    completed: data.completed === true,
    completedById: data.completedById ?? '',
    completedByName: data.completedByName ?? '',
    completedAt: data.completedAt?.toDate() ?? null,
    claimedById: data.claimedById ?? '',
    claimedByName: data.claimedByName ?? '',
    claimedAt: data.claimedAt?.toDate() ?? null,
    photosUploaded: data.photosUploaded === true,
  }
}

function mapNotice(data: TaskNoticeDoc | null | undefined): TaskNotice | null {
  if (!data) return null
  return {
    message: data.message ?? '',
    routeCode: data.routeCode ?? '',
    createdById: data.createdById ?? '',
    createdByName: data.createdByName ?? '',
    createdAt: data.createdAt?.toDate() ?? new Date(),
  }
}

function serializeRoute(route: TaskRoute): TaskRouteDoc {
  const payload: TaskRouteDoc = {
    routeCode: route.routeCode,
    note: route.note,
    completed: route.completed,
    completedById: route.completedById,
    completedByName: route.completedByName,
    completedAt: route.completedAt ? Timestamp.fromDate(route.completedAt) : null,
    claimedById: route.claimedById,
    claimedByName: route.claimedByName,
    claimedAt: route.claimedAt ? Timestamp.fromDate(route.claimedAt) : null,
    photosUploaded: route.photosUploaded,
  }
  if (isValidMapCoord(route.latitude, route.longitude)) {
    payload.latitude = route.latitude ?? undefined
    payload.longitude = route.longitude ?? undefined
  }
  return payload
}

function serializeNotice(notice: TaskNotice): TaskNoticeDoc {
  return {
    message: notice.message,
    routeCode: notice.routeCode,
    createdById: notice.createdById,
    createdByName: notice.createdByName,
    createdAt: Timestamp.fromDate(notice.createdAt),
  }
}

function mapTask(id: string, data: TaskDoc): Task {
  const latitude =
    typeof data.latitude === 'number' && Number.isFinite(data.latitude)
      ? data.latitude
      : null
  const longitude =
    typeof data.longitude === 'number' && Number.isFinite(data.longitude)
      ? data.longitude
      : null
  const routes = normalizeTaskRoutes({
    routes: (data.routes ?? []).map(mapRoute),
    routeCode: data.routeCode,
    latitude,
    longitude,
    status: data.status,
    completedById: data.completedById,
    completedByName: data.completedByName,
    completedAt: data.completedAt?.toDate() ?? null,
  })
  const primary = primaryTaskRoute(routes)
  return {
    id,
    title: data.title,
    description: data.description ?? '',
    status: data.status ?? Status.Pendiente,
    dueDate: data.dueDate?.toDate() ?? null,
    areaId: data.areaId ?? '',
    areaName: data.areaName ?? '',
    routeCode: primary?.routeCode ?? data.routeCode ?? '',
    latitude: primary?.latitude ?? (isValidMapCoord(latitude, longitude) ? latitude : null),
    longitude:
      primary?.longitude ?? (isValidMapCoord(latitude, longitude) ? longitude : null),
    routes,
    lastNotice: mapNotice(data.lastNotice),
    assignToAllTechnicians: data.assignToAllTechnicians === true,
    assignedTechnicianIds: data.assignedTechnicianIds ?? [],
    assignedTechnicianNames: data.assignedTechnicianNames ?? [],
    createdById: data.createdById,
    createdByName: data.createdByName,
    completedAt: data.completedAt?.toDate() ?? null,
    completedById: data.completedById ?? '',
    completedByName: data.completedByName ?? '',
    createdAt: data.createdAt.toDate(),
    updatedAt: data.updatedAt.toDate(),
  }
}

function applyCoords(
  payload: TaskDoc,
  latitude: number | null | undefined,
  longitude: number | null | undefined,
): void {
  if (isValidMapCoord(latitude, longitude)) {
    payload.latitude = latitude ?? undefined
    payload.longitude = longitude ?? undefined
    return
  }
  delete payload.latitude
  delete payload.longitude
}

function sortByUpdatedDesc(tasks: Task[]): Task[] {
  return [...tasks].sort(
    (left, right) => right.updatedAt.getTime() - left.updatedAt.getTime(),
  )
}

export class FirebaseTaskRepository implements TaskRepository {
  private readonly collectionRef = collection(firestoreDb, 'tasks')

  async getById(id: string): Promise<Task | null> {
    const snapshot = await getDoc(doc(this.collectionRef, id))
    if (!snapshot.exists()) return null
    return mapTask(snapshot.id, snapshot.data() as TaskDoc)
  }

  async listAll(): Promise<Task[]> {
    const snapshot = await getDocs(
      query(this.collectionRef, orderBy('updatedAt', 'desc')),
    )
    return snapshot.docs.map((item) =>
      mapTask(item.id, item.data() as TaskDoc),
    )
  }

  async listAccessibleForUser(userId: string): Promise<Task[]> {
    const [assignedSnap, allTechsSnap] = await Promise.all([
      getDocs(
        query(
          this.collectionRef,
          where('assignedTechnicianIds', 'array-contains', userId),
        ),
      ),
      getDocs(
        query(
          this.collectionRef,
          where('assignToAllTechnicians', '==', true),
        ),
      ),
    ])

    const byId = new Map<string, Task>()
    for (const item of [...assignedSnap.docs, ...allTechsSnap.docs]) {
      byId.set(item.id, mapTask(item.id, item.data() as TaskDoc))
    }
    return sortByUpdatedDesc([...byId.values()])
  }

  watchAll(
    onData: (tasks: Task[]) => void,
    onError?: (error: Error) => void,
  ): () => void {
    return onSnapshot(
      query(this.collectionRef, orderBy('updatedAt', 'desc')),
      (snapshot) => {
        onData(
          snapshot.docs.map((item) => mapTask(item.id, item.data() as TaskDoc)),
        )
      },
      (error) => onError?.(error),
    )
  }

  watchAccessibleForUser(
    userId: string,
    onData: (tasks: Task[]) => void,
    onError?: (error: Error) => void,
  ): () => void {
    let assigned: Task[] = []
    let allTechs: Task[] = []

    const emit = () => {
      const byId = new Map<string, Task>()
      for (const task of [...assigned, ...allTechs]) {
        byId.set(task.id, task)
      }
      onData(sortByUpdatedDesc([...byId.values()]))
    }

    const unsubAssigned = onSnapshot(
      query(
        this.collectionRef,
        where('assignedTechnicianIds', 'array-contains', userId),
      ),
      (snapshot) => {
        assigned = snapshot.docs.map((item) =>
          mapTask(item.id, item.data() as TaskDoc),
        )
        emit()
      },
      (error) => onError?.(error),
    )
    const unsubAll = onSnapshot(
      query(
        this.collectionRef,
        where('assignToAllTechnicians', '==', true),
      ),
      (snapshot) => {
        allTechs = snapshot.docs.map((item) =>
          mapTask(item.id, item.data() as TaskDoc),
        )
        emit()
      },
      (error) => onError?.(error),
    )

    return () => {
      unsubAssigned()
      unsubAll()
    }
  }

  async create(input: CreateTaskInput): Promise<Task> {
    const now = Timestamp.now()
    const routes = normalizeTaskRoutes({ routes: input.routes })
    const primary = primaryTaskRoute(routes)
    const payload: TaskDoc = {
      title: input.title,
      description: input.description,
      status: input.status,
      dueDate: input.dueDate ? Timestamp.fromDate(input.dueDate) : null,
      areaId: input.areaId,
      areaName: input.areaName,
      routeCode: primary?.routeCode ?? input.routeCode,
      routes: routes.map(serializeRoute),
      assignToAllTechnicians: input.assignToAllTechnicians,
      assignedTechnicianIds: input.assignedTechnicianIds,
      assignedTechnicianNames: input.assignedTechnicianNames,
      createdById: input.createdById,
      createdByName: input.createdByName,
      completedAt: null,
      completedById: '',
      completedByName: '',
      createdAt: now,
      updatedAt: now,
    }
    applyCoords(
      payload,
      primary?.latitude ?? input.latitude,
      primary?.longitude ?? input.longitude,
    )
    const id = crypto.randomUUID()
    await setDoc(doc(this.collectionRef, id), payload)
    return mapTask(id, payload)
  }

  async update(id: string, input: UpdateTaskInput): Promise<Task> {
    const ref = doc(this.collectionRef, id)
    const existing = await getDoc(ref)
    if (!existing.exists()) {
      throw new NotFoundError('Tarea no encontrada')
    }

    const current = existing.data() as TaskDoc
    const mapped = mapTask(id, current)
    const routes = normalizeTaskRoutes({
      routes: input.routes ?? mapped.routes,
      routeCode: input.routeCode ?? mapped.routeCode,
      latitude: input.latitude === undefined ? mapped.latitude : input.latitude,
      longitude: input.longitude === undefined ? mapped.longitude : input.longitude,
      status: input.status ?? mapped.status,
      completedById: input.completedById ?? mapped.completedById,
      completedByName: input.completedByName ?? mapped.completedByName,
      completedAt:
        input.completedAt === undefined ? mapped.completedAt : input.completedAt,
    })
    const primary = primaryTaskRoute(routes)
    const lastNotice =
      input.lastNotice === undefined ? mapped.lastNotice : input.lastNotice

    const payload: TaskDoc = {
      ...current,
      title: input.title ?? current.title,
      description: input.description ?? current.description,
      status: input.status ?? current.status,
      dueDate:
        input.dueDate === undefined
          ? current.dueDate
          : input.dueDate
            ? Timestamp.fromDate(input.dueDate)
            : null,
      areaId: input.areaId ?? current.areaId,
      areaName: input.areaName ?? current.areaName,
      routeCode: primary?.routeCode ?? current.routeCode,
      routes: routes.map(serializeRoute),
      assignToAllTechnicians:
        input.assignToAllTechnicians ?? current.assignToAllTechnicians,
      assignedTechnicianIds:
        input.assignedTechnicianIds ?? current.assignedTechnicianIds,
      assignedTechnicianNames:
        input.assignedTechnicianNames ?? current.assignedTechnicianNames,
      completedAt:
        input.completedAt === undefined
          ? current.completedAt
          : input.completedAt
            ? Timestamp.fromDate(input.completedAt)
            : null,
      completedById:
        input.completedById === undefined
          ? current.completedById
          : input.completedById,
      completedByName:
        input.completedByName === undefined
          ? current.completedByName
          : input.completedByName,
      updatedAt: Timestamp.now(),
    }

    if (lastNotice) {
      payload.lastNotice = serializeNotice(lastNotice)
    } else {
      payload.lastNotice = null
    }

    applyCoords(
      payload,
      primary?.latitude ?? null,
      primary?.longitude ?? null,
    )

    await updateDoc(ref, { ...payload })
    return mapTask(id, payload)
  }

  async delete(id: string): Promise<void> {
    await deleteDoc(doc(this.collectionRef, id))
  }
}
